A50-xds-outlier-detection.md

```diff
@@ -0,0 +1,207 @@
+  // double. That is, if the desired factor is 1.9, the runtime value should
+  // be 1900. Defaults to 1900.
+  google.protobuf.UInt32Value success_rate_stdev_factor = 5;
+
+  // The % chance that an address will be actually ejected when an outlier status
+  // is detected through success rate statistics. This setting can be used to
+  // disable ejection or to ramp it up slowly. Defaults to 100.
+  google.protobuf.UInt32Value enforcing_success_rate = 6;
+
+  // The number of addresses that must have enough request volume to
+  // detect success rate outliers. If the number of addresses is less than this
+  // setting, outlier detection via success rate statistics is not performed
+  // for any addresses. Defaults to 5.
+  google.protobuf.UInt32Value success_rate_minimum_hosts = 7;
+
+  // The minimum number of total requests that must be collected in one
+  // interval (as defined by the interval duration above) to include this address
+  // in success rate based outlier detection. If the volume is lower than this
+  // setting, outlier detection via success rate statistics is not performed
+  // for that address. Defaults to 100.
+  google.protobuf.UInt32Value success_rate_request_volume = 8;
+
+  // The failure percentage to use when determining failure percentage-based outlier detection. If
+  // the failure percentage of a given address is greater than or equal to this value, it will be
+  // ejected. Defaults to 85.
+  google.protobuf.UInt32Value failure_percentage_threshold = 9;
+
+  // The % chance that an address will be actually ejected when an outlier status is detected through
+  // failure percentage statistics. This setting can be used to disable ejection or to ramp it up
+  // slowly. Defaults to 0.
+  google.protobuf.UInt32Value enforcing_failure_percentage = 10;
+
+  // The minimum number of addresses in order to perform failure percentage-based ejection.
+  // If the total number of addresses is less than this value, failure percentage-based
+  // ejection will not be performed. Defaults to 5.
+  google.protobuf.UInt32Value failure_percentage_minimum_hosts = 11;
+
+  // The minimum number of total requests that must be collected in one interval (as defined by the
+  // interval duration above) to perform failure percentage-based ejection for this address. If the
+  // volume is lower than this setting, failure percentage-based ejection will not be performed for
+  // this host. Defaults to 50.
+  google.protobuf.UInt32Value failure_percentage_request_volume = 12;
+
+  // The config for the child policy
+  repeated LoadBalancingConfig child_policy = 13;
+}
+```
+
+### Outlier Detection Behavior
+
+Internally, the `outlier_detection` LB policy will store a map with keys that are subchannel addresses, and values that are objects that contain the following:
+
+ - The call result counter object (see [Call Counter section](#call-counter))
+ - The latest ejection timestamp, or `null` if the address is currently not ejected
```

Comment:
Could just set the timestamp to a representation of infinite-future.  In C-core, I would expect that to be fairly idiomatic, and it would not require using `absl::optional<>`.

Reply:
The algorithm does not just check the ejection time relative to the current time, it also has different behavior depending on whether an address currently is ejected or not. I found it useful to have a sentinel value to check for that, and having that be `null` clearly distinguishes it from a valid timestamp. Alternatively, we could have an additional boolean `ejected` field, and ignore this timestamp if it is false.

Reply:
In other words, I think that the behavior requires logic that is equivalent to interacting with an `optional` anyway, and I think that actually having an `optional` is a good way to express it.

Reply:
Okay, I'll buy that argument.